src/NITTA/Intermediate/Analysis.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- |
+Module      : NITTA.Intermediate.Analysis
+Description : Analysis of the process execution flow
+Copyright   : (c) Aleksandr Penskoi, 2022
+License     : BSD3
+Maintainer  : aleksandr.penskoi@gmail.com
+Stability   : experimental
+
+Much of the work in this module focuses on building waves of process execution.
+In this case, we call a wave a set of functions that are ready for execution (all input vars are ready to use)
+and that can be performed independently of each other.
+Example:
+
+Lua code:
+
+function sum(a)
+    local d = a + 1
+    sum(d)
+end
+sum(0)
+
+After building DataFlowGraph, we get the following set of functions:
+- const(1) = !1#0
+- loop(0, d^0#0) = a^0#0
+- a^0#0 + !1#0 = d^0#0
+
+Const and Loop function can be executed at the first wave because all input variables are ready to use.
+After executing first wave a^0#0 and !1#0 will be ready, so we will be able to execute add function.
+So the resulting process waves are the following:
+[
+    ProcessWave {
+        pwFs = [const(1) = !1#0,loop(0, d^0#0) = a^0#0],
+        pwOut = fromList ["!1#0","a^0#0"]
+    },
+    ProcessWave {
+        pwFs = [a^0#0 + !1#0 = d^0#0],
+        pwOut = fromList ["d^0#0"]
+    }
+]
+-}
```

Comment:
Use haddock syntax. It will be a part of documentation.

Reply:
I've added markup for code sections
![image](https://user-images.githubusercontent.com/31983783/162616430-20b6de86-2e34-4d58-9184-175c00862bc4.png)

Reply:
nice